concurrency/DFSClient.java

```diff
@@ -0,0 +1,2910 @@
+    checkOpen();
+    try (TraceScope ignored = tracer.newScope("addCacheDirective")) {
+      return namenode.addCacheDirective(info, flags);
+    } catch (RemoteException re) {
+      throw re.unwrapRemoteException();
+    }
+  }
+
+  public void modifyCacheDirective(
+      CacheDirectiveInfo info, EnumSet<CacheFlag> flags) throws IOException {
+    checkOpen();
+    try (TraceScope ignored = tracer.newScope("modifyCacheDirective")) {
+      namenode.modifyCacheDirective(info, flags);
+    } catch (RemoteException re) {
+      throw re.unwrapRemoteException();
+    }
+  }
+
+  public void removeCacheDirective(long id)
+      throws IOException {
+    checkOpen();
+    try (TraceScope ignored = tracer.newScope("removeCacheDirective")) {
+      namenode.removeCacheDirective(id);
+    } catch (RemoteException re) {
+      throw re.unwrapRemoteException();
+    }
+  }
+
+  public RemoteIterator<CacheDirectiveEntry> listCacheDirectives(
+      CacheDirectiveInfo filter) throws IOException {
+    return new CacheDirectiveIterator(namenode, filter, tracer);
+  }
+
+  public void addCachePool(CachePoolInfo info) throws IOException {
+    checkOpen();
+    try (TraceScope ignored = tracer.newScope("addCachePool")) {
+      namenode.addCachePool(info);
+    } catch (RemoteException re) {
+      throw re.unwrapRemoteException();
+    }
+  }
+
+  public void modifyCachePool(CachePoolInfo info) throws IOException {
+    checkOpen();
+    try (TraceScope ignored = tracer.newScope("modifyCachePool")) {
+      namenode.modifyCachePool(info);
+    } catch (RemoteException re) {
+      throw re.unwrapRemoteException();
+    }
+  }
+
+  public void removeCachePool(String poolName) throws IOException {
+    checkOpen();
+    try (TraceScope ignored = tracer.newScope("removeCachePool")) {
+      namenode.removeCachePool(poolName);
+    } catch (RemoteException re) {
+      throw re.unwrapRemoteException();
+    }
+  }
+
+  public RemoteIterator<CachePoolEntry> listCachePools() throws IOException {
+    return new CachePoolIterator(namenode, tracer);
+  }
+
+  /**
+   * Save namespace image.
+   *
+   */
+  void saveNamespace() throws IOException {
+    try (TraceScope ignored = tracer.newScope("saveNamespace")) {
+      namenode.saveNamespace();
+    } catch (RemoteException re) {
+      throw re.unwrapRemoteException(AccessControlException.class);
+    }
+  }
+
+  /**
+   * Rolls the edit log on the active NameNode.
+   * @return the txid of the new log segment
+   *
+   * @see ClientProtocol#rollEdits()
+   */
+  long rollEdits() throws IOException {
+    try (TraceScope ignored = tracer.newScope("rollEdits")) {
+      return namenode.rollEdits();
+    } catch (RemoteException re) {
+      throw re.unwrapRemoteException(AccessControlException.class);
+    }
+  }
+
+  @VisibleForTesting
+  ExtendedBlock getPreviousBlock(long fileId) {
+    return filesBeingWritten.get(fileId).getBlock();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **2265**. At this line the following field is not protected by synchronization: `filesBeingWritten`. Your code typically protects this field with synchronization on object: `filesBeingWritten`.

**Fix**  
 Consider protecting the above field with synchronization on object: `filesBeingWritten`. For example, like at lines: **439**, **429**, **430**, **454**, **402**, **503**, **507**, and **506**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **402**, **416**, **429**, **430**, **439**, **454**, **503**, **506**, and **507**. This is a call to `get()`. You are already protecting `isEmpty()`. You are protecting these operations at lines: **439**.